### PR TITLE
Close #1145: fix bug in custom backward of least square solver for non-leaf tensor

### LINKED ISSUE #1145: Bug in implicit backward solver for non-leaf tensors
### Environment

latest

### Describe the bug

The custom backward of least square solvers currently give the incorrect gradient w.r.t the physics parameters, if they are non-leaf tensors. 

### Steps to reproduce

This can happen when, for example we train a neural network to estimate the physics parameters, then do back-propagation though a reconstruction net (eg unrolled, RAM) that uses least square solver:

```python

# %%
import torch
import deepinv as dinv

device = "cuda"
torch.manual_seed(0)

filter = dinv.physics.blur.gaussian_blur(sigma=(4.5, 4.5), device=device)

physics = dinv.physics.Blur(filter=filter, device=device, max_iter=200)
rec_model = dinv.models.RAM(pretrained=True, device=device)
for param in rec_model.parameters():
    param.requires_grad = False
rec_model.eval()


# Train a kernel estimation model
kernel_model = dinv.models.KernelIdentificationNetwork(
    pretrained="download", device=device
)
for p in kernel_model.parameters():
    p.grad = None

x = torch.rand(1, 1, 256, 256, device=device)

# Keep the same measurement for both branches to compare like-for-like gradients.
with torch.no_grad():
    y_fixed = physics(x).detach()


# %%
physics.implicit_backward_solver = False
params = kernel_model(y_fixed)
filter_hat = params["filters"][:, :, 0]
physics.update_parameters(filter=filter_hat)
x_hat = rec_model(y_fixed, physics=physics, sigma=0.01)

physics.filter.retain_grad()  # retain grad for filter to check against implicit backward

loss = (x_hat - x).pow(2).mean()
loss.backward()
true_grad = physics.filter.grad

# %%
physics.implicit_backward_solver = True
physics.filter = physics.filter.detach().clone()
physics.filter.grad = None

for p in kernel_model.parameters():
    p.grad = None
for p in rec_model.parameters():
    p.grad = None
params = kernel_model(y_fixed)
filter_hat = params["filters"][:, :, 0]
physics.update_parameters(filter=filter_hat)
x_hat = rec_model(y_fixed, physics=physics, sigma=0.01)
loss = (x_hat - x).pow(2).mean()
loss.backward()
implicit_grad = physics.filter.grad

# %%
print("True grad:", true_grad)
print("Implicit grad:", implicit_grad)
diff = implicit_grad - true_grad
rel_err = diff.norm() / (true_grad.norm() + 1e-12)
cos = torch.nn.functional.cosine_similarity(
    implicit_grad.flatten(), true_grad.flatten(), dim=0
)
print("Relative error:", rel_err.item())
print("Cosine similarity:", cos.item())

```
This give 

```
True grad: tensor([[[[0.5386, 0.5450, 0.4921,  ..., 0.3816, 0.4019, 0.4210],
          [0.4746, 0.4809, 0.4281,  ..., 0.3166, 0.3370, 0.3563],
          [0.3706, 0.3769, 0.3237,  ..., 0.2099, 0.2306, 0.2502],
          ...,
          [0.3667, 0.3686, 0.3119,  ..., 0.2326, 0.2548, 0.2727],
          [0.4330, 0.4350, 0.3788,  ..., 0.2998, 0.3220, 0.3397],
          [0.4514, 0.4531, 0.3968,  ..., 0.3184, 0.3405, 0.3581]]]],
       device='cuda:0')
Implicit grad: tensor([[[[-3.9124e-02, -1.4905e-03, -2.2849e-02,  ..., -7.5977e-03,
           -9.8136e-03, -8.8548e-03],
          [-3.0722e-02,  6.9987e-03, -1.4152e-02,  ...,  1.5290e-03,
           -8.2322e-04,  2.2074e-05],
          [-4.9971e-02, -1.2088e-02, -3.3668e-02,  ..., -2.0591e-02,
           -2.2791e-02, -2.1940e-02],
          ...,
          [-3.4220e-02,  8.3899e-04, -2.4114e-02,  ..., -7.9989e-03,
           -9.5070e-03, -1.1387e-02],
          [-1.3807e-02,  2.1234e-02, -3.3321e-03,  ...,  1.1570e-02,
            1.0053e-02,  8.1890e-03],
          [-3.9485e-02, -4.9428e-03, -2.9637e-02,  ..., -1.2933e-02,
           -1.4372e-02, -1.6157e-02]]]], device='cuda:0')
Relative error: 1.0545765161514282
Cosine similarity: -0.24902009963989258
```

This is due to a wrong manipulation of grad of non-leaf tensors. 

### Expected behavior

The `implicit_grad` and `true_grad` should be close

### Full traceback (if applicable)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mh-nguyen712**
This PR fixes #1145 by properly handle the gradient for non-leaf tensors, in the custom backward of the least squares solvers. 

Thank you for contributing to DeepInverse!

Please refer to our [contributing guidelines](https://deepinv.github.io/deepinv/contributing.html) for full instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/docs_cpu.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

File: deepinv/tests/test_optim.py [hunk 1]
```diff
@@ -1262,6 +1262,77 @@ def test_least_squares_implicit_backward(device, solver, physics_name, batch_siz
...
+    x_hat = physics.prox_l2(z, y, solver="CG", tol=1e-6, max_iter=50, gamma=1e-3)
+    loss = (x_hat - x).pow(2).mean()
+    loss.backward()
+    implicit_grad = physics.mat.grad.detach().clone()
+
+    torch.testing.assert_close(
+        implicit_grad,
+        explicit_grad,
+        rtol=1e-2,
+        atol=1e-2,
```
Comment (Tmodrzyk) [inline deepinv/tests/test_optim.py]: I don't know the expected range of numerical errors in this case but these bounds seem large.
Shouldn't we tighten them to make sure the gradients are indeed equal ?

File: deepinv/tests/test_optim.py [hunk 2]
```diff
@@ -1262,6 +1262,77 @@ def test_least_squares_implicit_backward(device, solver, physics_name, batch_siz
     torch.use_deterministic_algorithms(prev_deterministic)
 
 
+def test_least_squares_implicit_backward_nonleaf_buffer_grad(device):
+    """Compare explicit vs implicit gradient on an intermediate non-leaf physics buffer."""
```
Comment (Tmodrzyk) [inline deepinv/tests/test_optim.py:1266]: This tests adds quite a lot of lines to the test file, for a fairly specific test case.
Could we integrate it to the ``test_least_squares_implicit_backward`` test ?

File: deepinv/tests/test_optim.py [hunk 3]
```diff
@@ -1262,6 +1262,77 @@ def test_least_squares_implicit_backward(device, solver, physics_name, batch_siz
...
+    class DummyPhysics(dinv.physics.LinearPhysics):
+        def __init__(self, mat):
+            super().__init__()
+            self.register_buffer("mat", mat)
+
+        def A(self, x):
+            return x @ self.mat.transpose(0, 1)
+
+        def A_adjoint(self, y):
+            return y @ self.mat
```
Comment (Tmodrzyk) [inline deepinv/tests/test_optim.py:1282]: We already have another ``DummyPhysics`` implemented in this file (in `test_condition_number`).
It would be awesome to have the same one in both tests (if possible), or to move them to ``dummy.py``.

**Review by Tmodrzyk — COMMENTED** [review]
Thanks a lot for this bugfix @mh-nguyen712 !
The fix lgtm, but I have a few comments to make the test more compact, as the test is fairly specific and the ``test_optim.py`` file is growing larger.  
Edit: We should also update the ``changelog.rst`` + add a PR description :smile:

File: deepinv/tests/test_optim.py (hunk 1 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_optim.py]: Changed to 1e-4

File: deepinv/tests/test_optim.py (hunk 3 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_optim.py:1282]: I would need a physics which has parameter (buffer, here is `mat`) that will be optimized. I think it's quite specific use-case. 
I also need the DummyModel has parameters to differentiate (different than the one in dummy.py, I don't want to break other tests)

File: deepinv/tests/test_optim.py (hunk 2 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_optim.py:1266]: Hmm this one involves a neural net, so integrate to `test_least_squares_implicit_backward` would make the test more longer. I'm not sure this is optimal neither.

**Comment by Tmodrzyk** [comment]
Hm my workarounds for making the test shorter were not really applicable, I think we can leave it as is for now ! 
I'll merge once the tests are passing. :smile:

**Comment by mh-nguyen712** [comment]
Thanks for the review @Tmodrzyk ! It seems that there is something wrong with /gpu-tests
```Run pixi run -e full python -c "import deepinv"
Error:   × failed to link urllib3-2.6.3-pyhd8ed1ab_0.conda
  ├─▶ failed to link 'site-packages/urllib3/contrib/emscripten/fetch.py'
  ├─▶ failed to copy file to destination
  ╰─▶ failed to copy file from /local/jtachell/.cache/rattler/cache/pkgs/urllib3-2.6.3-pyhd8ed1ab_0/site-packages/urllib3/contrib/emscripten/fetch.py to /home/jtachell/pixi-envs/deepinv-10826435223575736850/envs/full/lib/python3.14/site-packages/urllib3/contrib/emscripten/fetch.py: Quota exceeded (os error 122)
  ````

**Comment by Tmodrzyk** [comment]
Indeed, we better wait for the hotfix to be merged then try again the GPU tests before merging this PR. :smile:

**Review by Tmodrzyk — APPROVED** [review]
LGTM
